python/lsst/ctrl/bps/generic_workflow.py

```diff
@@ -581,3 +638,102 @@ def validate(self):
+        self.add_edges_from(workflow.edges())
+        for source in self_sources:
+            for sink in new_sinks:
+                self.add_edge(sink, source)
+
+        # Files are stored separately so copy them.
+        for job_name in workflow:
+            self.add_job_inputs(job_name, workflow.get_job_inputs(job_name, data=True))
+            self.add_job_outputs(job_name, workflow.get_job_outputs(job_name, data=True))
+            self.add_executable(workflow.get_job(job_name).executable)
+
+    def add_final(self, final):
+        """Add special final job/workflow to the generic workflow.
+
+        Parameters
+        ----------
+        final : `lsst.ctrl.bps.GenericWorkflowJob` or
+                `lsst.ctrl.bps.GenericWorkflow`
+            Information needed to execute the special final job(s), the
+            job(s) to be executed after all jobs that can be executed
+            have been executed regardless of exit status of any of the
+            jobs.
+        """
+        if not isinstance(final, GenericWorkflowJob) and not isinstance(final, GenericWorkflow):
+            raise TypeError("Invalid type for GenericWorkflow final ({type(final)})")
+
+        self._final = final
+        if isinstance(final, GenericWorkflowJob):
+            self.add_executable(final.executable)
+
+    def get_final(self):
+        """Return job/workflow to be executed after all jobs that can be
+        executed have been executed regardless of exit status of any of
+        the jobs.
+
+        Returns
+        -------
+        final : `lsst.ctrl.bps.GenericWorkflowJob` or
+                `lsst.ctrl.bps.GenericWorkflow`
+            Information needed to execute final job(s).
+        """
+        return self._final
+
+    def add_executable(self, executable):
+        """Add executable to workflow's list of executables.
+
+        Parameters
+        ----------
+        executable : `lsst.ctrl.bps.GenericWorkflowExec`
+            Executable object to be added to workflow.
+        """
+        if executable is not None:
+            self._executables[executable.name] = executable
+        else:
+            _LOG.warning("add_executable the executable is None")
+
+    def get_executables(self, data=False, transfer_only=True):
+        """Retrieve executables from generic workflow.
+
+        Parameters
+        ----------
+        data : `bool`, optional
+            Whether to return the executable data as well as the exec object
+            name.  (The defaults is False.)
+        transfer_only : `bool`, optional
+            Whether to only return executables for which transfer_executable
+            is True.
+
+        Returns
+        -------
+        execs : `list` [`lsst.ctrl.bps.GenericWorkflowExec`] or `list` [`str`]
+            Filtered executable names or objects from generic workflow.
+        """
+        execs = []
+        for name, executable in self._execs.items():
```

Comment:
Unless I'm missing something, it doesn't look like the **GenericWorkflow** has the `_execs` attribute. Did you mean `self._executables`?